src/server/tran_server.cpp

```diff
@@ -487,55 +538,93 @@ tran_server::connection_handler::get_request_handlers ()
 void
 tran_server::connection_handler::receive_disconnect_request (page_server_conn_t::sequenced_payload &&)
 {
-  m_is_disconnecting.store (true);
-  m_conn->stop_response_broker (); // wake up threads waiting for a response and tell them it won't be served.
-  m_ts.disconnect_page_server_async (this);
+  constexpr bool with_disconnect_msg = true;
+  disconnect_async (with_disconnect_msg);
 }
 
 void
-tran_server::connection_handler::push_request (tran_to_page_request reqid, std::string &&payload)
+tran_server::connection_handler::push_request_internal (tran_to_page_request reqid, std::string &&payload)
 {
+  // make sure the m_conn is not nullptr;
   m_conn->push (reqid, std::move (payload));
 }
 
 int
-tran_server::connection_handler::send_receive (tran_to_page_request reqid, std::string &&payload_in,
-    std::string &payload_out) const
+tran_server::connection_handler::push_request (tran_to_page_request reqid, std::string &&payload)
 {
-  const css_error_code error_code = m_conn->send_recv (reqid, std::move (payload_in), payload_out);
-  // NOTE: enhance error handling when:
-  //  - more than one page server will be handled
-  //  - fail-over will be implemented
+  auto slock_state = std::shared_lock<std::shared_mutex> { m_state_mtx };
+
+  if (m_state != state::CONNECTED)
+    {
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_CONN_PAGE_SERVER_CANNOT_BE_REACHED, 0);
+      return ER_CONN_PAGE_SERVER_CANNOT_BE_REACHED;
+    }
+
+  // state::CONNECTED guarantees that the internal m_node is not nullptr.
+  auto slock_conn = std::shared_lock<std::shared_mutex> { m_conn_mtx };
+  assert (m_conn != nullptr); // because m_conn is set with both m_state_mtx and m_conn_mtx locked.
+  slock_state.unlock ();
```

Comment:
```suggestion
  slock_conn.unlock ();
  slock_state.unlock ();
```
I don't think that without there is some inconsistency, but just so it "looks" nice. :)
Same in `send_receive`.

Reply:
`slock_state` must be unlocked to allow `disconnect_async` which also locks `m_state_mtx`. It can be called by the request handler, and if this `send_receive` is sleeping with the mutex locked, a deadlock will happen.
```
client A:                       send_receive(m_state_mtx locked) and sleep
the request handler:   
   - SEND_DISCONNECT_REQUEST_MSG: disconenct_async(): blocked by m_state_mtx
   - RESPONSE: serve to client A (will wake up client A, but can't server this for good)
```
And `slock_conn` must be locked, or `m_node` can be set to `nullptr` by disconnecting.

Reply:
Actually, this is about `send_receive`, not about `push_request`. In the case of `push_request`, it is just for code consistency.

Reply:
For both of them, the `m_conn_mtx` is needed.
I overlooked the call to `*_internal` functions when I reviewed. Sorry.
- for `push_request` the code translates to :
```
  auto slock_conn = std::shared_lock<std::shared_mutex> { m_conn_mtx };
  assert (m_conn != nullptr); // because m_conn is set with both m_state_mtx and m_conn_mtx locked.
  slock_state.unlock ();

  //push_request_internal (reqid, std::move (payload));
  // make sure the m_conn is not nullptr;
  m_conn->push (reqid, std::move (payload)); // <<------------- slock_conn needed here

```
- for `send_receive` the code transalates to:
```
  auto slock_conn = std::shared_lock<std::shared_mutex> { m_conn_mtx };
  assert (m_conn != nullptr); // because m_conn is set with both m_state_mtx and m_conn_mtx locked.
  slock_state.unlock (); // to allow disconnect_async ()

  //return send_receive_internal (reqid, std::move (payload_in), payload_out);
  // make sure the m_conn is not nullptr;
  // <<------------------ slock_conn needed from here on
  const css_error_code error_code = m_conn->send_recv (reqid, std::move (payload_in), payload_out);
  if (error_code != NO_ERRORS)
    {
      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_CONN_PAGE_SERVER_CANNOT_BE_REACHED, 0);
      return ER_CONN_PAGE_SERVER_CANNOT_BE_REACHED;
    }

  return NO_ERROR;

```